File: dbassistant/layers/PromptPreparationLayer.py
import re
from typing import List
from dbassistant.interfaces.ILayer import ILayer
from dbassistant.layers.LayerCargo import LayerCargo
from dbassistant.tokens.Tokenizer import Tokenizer


class PromptPreparationLayer(ILayer):
    def __init__(self, prompt_size: int):
        self.prompt_size = prompt_size

    def process(self, cargo: LayerCargo) -> LayerCargo:
        input_string = cargo.take()
        assert isinstance(input_string, str)
        special_character = re.compile(r'[^A-z0-9 _@]')

        spaced = ""
        for char in input_string:
            if special_character.match(char):
                spaced += " " + char + " "
            else:
                spaced += char
        spaced = re.sub(r' +', ' ', spaced)
        symbols = spaced.split(" ")
        # filter empty strings
        symbols = list(filter(None, symbols))
        symbols.append("[END]")

        cargo.put(symbols)
        return cargo

File: dbassistant/layers/PromptPreparationLayer.ts
import ILayer from '../interfaces/ILayer';
import LayerCargo from '../layers/LayerCargo';
import Tokenizer from '../tokens/Tokenizer';

export default class PromptPreparationLayer implements ILayer {
    private promptSize: number;

    constructor(promptSize: number) {
        this.promptSize = promptSize;
    }

    process(cargo: LayerCargo): LayerCargo {
        const inputString = cargo.take();
        if (typeof inputString !== 'string') {
            throw new Error('Input cargo should be a string.');
        }

        const specialCharacter = /[^A-z0-9 _@]/;
        let spaced = "";
        for (const char of inputString) {
            if (specialCharacter.test(char)) {
                spaced += " " + char + " ";
            } else {
                spaced += char;
            }
        }
        spaced = spaced.replace(/ +/g, ' ');
        const symbols = spaced.split(" ").filter(symbol => symbol !== "");

        symbols.push("[END]");

        cargo.put(symbols);
        return cargo;
    }
}

